### PR TITLE
refactor: check security

### PR DESCRIPTION
**Pull request by Randolph314**
## Description

Closes [#1502](https://mask.atlassian.net/browse/MF-1502)

## Type of change

- [ ] Documentation
- [x] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/shared-base/src/Messages/Mask.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ import type { SerializableTypedMessages } from '@masknet/typed-message'
 import type { ProfileIdentifier, PersonaIdentifier } from '../Identifier'
 import type { RelationFavor } from '../Persona/type'
 import type { EnhanceableSite, ExtensionSite } from '../Site'
+import type { ChainId } from '@masknet/web3-shared-evm'
```
Comment (Jack-Works) [inline packages/shared-base/src/Messages/Mask.ts]: circular dependency? shared-base should not refer to it. maybe you should move the definition of ChainId to this module.

File: packages/plugins/GoPlusSecurity/src/index.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import { registerPlugin } from '@masknet/plugin-infra'
 import { base } from './base'
+export * from './messages'
```
Comment (Jack-Works) [inline packages/plugins/GoPlusSecurity/src/index.ts]: DO NOT export from the `index.ts`. `index.ts` of a plugin should only provide the register of a plugin. Please import like `@pakcage/name/messages` and define it in the `package.json` using `exports` field and `typeVersions`.

File: packages/mask/src/social-network/ui.ts [hunk 3]
```diff
@@ -153,6 +153,7 @@ export async function activateSocialNetworkUIInner(ui_deferred: SocialNetworkUI.
                     },
                     getPersonaAvatar: Services.Identity.getPersonaAvatar,
                     ownProofChanged: MaskMessages.events.ownProofChanged,
+                    setPluginEnable: Services.Settings.setPluginMinimalModeEnabled,
```
Comment (Jack-Works) [inline packages/mask/src/social-network/ui.ts]: you should reverse the option.

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/shared-base/package.json [hunk 4]
```diff
@@ -13,6 +13,7 @@
   "dependencies": {
     "@masknet/public-api": "workspace:*",
     "@masknet/typed-message": "workspace:*",
+    "@masknet/web3-shared-evm": "workspace:*",
```
Comment (Jack-Works) [inline packages/shared-base/package.json]: Don't. Circular reference.

File: packages/plugin-infra/src/types.ts [hunk 5]
```diff
@@ -365,6 +365,7 @@ export namespace Plugin.SNSAdaptor {
         privileged_silentSign: () => (signer: ECKeyIdentifier, message: string) => Promise<PersonaSignResult>
         getPersonaAvatar: (identifier: ECKeyIdentifier | null | undefined) => Promise<string | null | undefined>
         ownProofChanged: UnboundedRegistry<void>
+        setPluginEnable: (id: string, enabled: boolean) => Promise<void>
```
Comment (Jack-Works) [inline packages/plugin-infra/src/types.ts]: we have another concept of enabling/disabling which is not exposed to users. Do you mean: `setPluginMinimalMode`

File: packages/mask/src/social-network/ui.ts (hunk 3 above)
Comment (Jack-Works) [inline packages/mask/src/social-network/ui.ts]:
```suggestion
                    setMinimalMode: Services.Settings.setPluginMinimalModeEnabled,
```

File: packages/plugins/GoPlusSecurity/src/SNSAdaptor/components/SecurityPanel.tsx [hunk 6]
```diff
@@ -71,14 +70,11 @@ export const SecurityPanel = memo<TokenCardProps>(({ tokenSecurity, tokenInfo, t
...
-        const makeMessageList = getMessageList(tokenSecurity)
-
-        const riskyFactors = makeMessageList.filter((x) => x.level === SecurityMessageLevel.High).length
-        const attentionFactors = makeMessageList.filter((x) => x.level === SecurityMessageLevel.Medium).length
-        return { riskyFactors, attentionFactors, makeMessageList }
-    }, [tokenSecurity])
+    const {
+        risk_item_quantity: riskyFactors = 0,
+        warn_item_quantity: attentionFactors = 0,
+        message_list: makeMessageList = [],
```
Comment (guanbinrui) [inline packages/plugins/GoPlusSecurity/src/SNSAdaptor/components/SecurityPanel.tsx]:
```suggestion
        message_list: makeMessageList = EMPTY_LIST,
```

File: packages/shared-base/src/Messages/Mask.ts [hunk 7]
```diff
@@ -10,6 +10,69 @@ enum NetworkPluginID {
     PLUGIN_SOLANA = 'com.mask.solana',
 }
 
+// Learn more about ethereum ChainId https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md
+export enum ChainId {
```
Comment (guanbinrui) [inline packages/shared-base/src/Messages/Mask.ts]: Please avoid doing this. Just use `number`.

File: packages/web3-providers/src/gopluslabs/index.ts [hunk 8]
```diff
@@ -41,3 +42,45 @@ export class GoPlusLabsAPI implements SecurityAPI.Provider<ChainId> {
...
+
+export const isHighRisk = (tokenSecurity?: SecurityAPI.TokenSecurityType) => {
+    if (!tokenSecurity) return false
+    return tokenSecurity.trust_list === '1'
+        ? false
+        : SecurityMessages.filter(
+              (x) =>
+                  x.condition(tokenSecurity) && x.level !== SecurityMessageLevel.Safe && !x.shouldHide(tokenSecurity),
+          )
+              .sort((a) => (a.level === SecurityMessageLevel.High ? -1 : 1))
```
Comment (guanbinrui) [inline packages/web3-providers/src/gopluslabs/index.ts]:
```suggestion
              .sort((a, z) => {
                  if (a.level === SecurityMessageLevel.High) return -1
                  if (z.level === SecurityMessageLevel.High) return 1
                  return 0
              })
```

File: packages/web3-providers/src/gopluslabs/index.ts [hunk 9]
```diff
@@ -41,3 +42,45 @@ export class GoPlusLabsAPI implements SecurityAPI.Provider<ChainId> {
...
+              .filter((x) => x.level === SecurityMessageLevel.High).length > 0
```
Comment (guanbinrui) [inline packages/web3-providers/src/gopluslabs/index.ts]: Merge this into the first `.filter()` callback.

File: packages/web3-providers/src/gopluslabs/index.ts [hunk 10]
```diff
@@ -41,3 +42,45 @@ export class GoPlusLabsAPI implements SecurityAPI.Provider<ChainId> {
...
+}
+
+export const getMessageList = (tokenSecurity: SecurityAPI.TokenSecurityType) =>
+    tokenSecurity.trust_list === '1'
+        ? []
+        : SecurityMessages.filter(
+              (x) =>
+                  x.condition(tokenSecurity) && x.level !== SecurityMessageLevel.Safe && !x.shouldHide(tokenSecurity),
+          ).sort((a) => (a.level === SecurityMessageLevel.High ? -1 : 1))
```
Comment (guanbinrui) [inline packages/web3-providers/src/gopluslabs/index.ts]:
```suggestion
          ).sort((a, z) => {
              if (a.level === SecurityMessageLevel.High) return -1
              if (z.level === SecurityMessageLevel.High) return 1
              return 0
          })
```

File: packages/web3-providers/src/gopluslabs/index.ts [hunk 11]
```diff
@@ -41,3 +42,45 @@ export class GoPlusLabsAPI implements SecurityAPI.Provider<ChainId> {
         return result.map((x) => ({ chainId: parseInt(x.id) ?? ChainId.Mainnet, name: x.name }))
     }
 }
+
+export const createTokenSecurity = (
+    chainId: ChainId,
+    response: Record<string, SecurityAPI.ContractSecurity & SecurityAPI.TokenSecurity & SecurityAPI.TradingSecurity>,
+) => {
+    response ??= {}
```
Comment (guanbinrui) [inline packages/web3-providers/src/gopluslabs/index.ts]:
```suggestion
    response: Record<string, SecurityAPI.ContractSecurity & SecurityAPI.TokenSecurity & SecurityAPI.TradingSecurity> = {},
) => {
```

File: packages/web3-providers/src/gopluslabs/index.ts [hunk 12]
```diff
@@ -41,3 +42,45 @@ export class GoPlusLabsAPI implements SecurityAPI.Provider<ChainId> {
...
+    if (isEmpty(response)) return
+    const entity = first(Object.entries(response ?? {}))
```
Comment (guanbinrui) [inline packages/web3-providers/src/gopluslabs/index.ts]:
```suggestion
    const entity = first(Object.entries(response))
```

File: packages/web3-providers/src/gopluslabs/constants.tsx [hunk 13]
```diff
@@ -0,0 +1,76 @@
+export const GO_PLUS_LABS_ROOT_URL = 'https://api.gopluslabs.io'
```
Comment (guanbinrui) [inline packages/web3-providers/src/gopluslabs/constants.tsx]: This file should ends with extension `.ts`.

File: packages/shared-base/src/Messages/Mask.ts [hunk 14]
```diff
@@ -101,6 +164,17 @@ export interface Web3ProfileDialogRequest {
     open: boolean
 }
 
+export interface CheckSecurityCloseConfirmDialogRequest {
+    open: boolean
+}
+
+export interface CheckSecurityDialogRequest {
```
Comment (guanbinrui) [inline packages/shared-base/src/Messages/Mask.ts]:
```
export type = {
    open: true
    searchHidden: boolean
    tokenAddress?: string
    chainId?: ChainId
} | {
    open: false
}
```

File: packages/plugins/GoPlusSecurity/src/messages.ts [hunk 15]
```diff
@@ -1,10 +1,19 @@
...
+interface Token {
+    contract: string
+    name: string
+    chainId: ChainId
+}
+
 type CheckSecurityDialogEvent = { open: boolean }
 
-interface PluginGoPlusSecurityMessage {
+type tokenRiskWarningDialogEvent = { open: boolean; token?: Token; swap: boolean }
```
Comment (guanbinrui) [inline packages/plugins/GoPlusSecurity/src/messages.ts]:
```ts
type tokenRiskWarningDialogEvent = {
    open: true
    token?: Token
    swap: boolean
} | {
    open: false
}
```

File: packages/plugins/GoPlusSecurity/src/SNSAdaptor/components/RiskWarningDialog.tsx [hunk 16]
```diff
@@ -54,74 +56,87 @@ const useStyles = makeStyles()((theme) => ({
...
                             onClick={onCopy}>
                             <Copy size={14} />
                         </Link>
                         <Link
                             className={classes.link}
                             href={
                                 explorerResolver.fungibleTokenLink?.(
-                                    tokenInfo?.chainId ?? 1,
-                                    tokenInfo?.contract ?? ZERO_ADDRESS,
+                                    token?.chainId ?? 1,
```
Comment (guanbinrui) [inline packages/plugins/GoPlusSecurity/src/SNSAdaptor/components/RiskWarningDialog.tsx]:
```suggestion
                                    token?.chainId ?? ChainId.Mainnet,
```

File: packages/shared-base/src/Messages/Mask.ts (hunk 7 above)
Comment (Jack-Works) [reply packages/shared-base/src/Messages/Mask.ts]: why? 🤔

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**